src/core/components/pin/gc.js

```diff
@@ -0,0 +1,153 @@
+'use strict'
+
+const promisify = require('promisify-es6')
+const CID = require('cids')
+const base32 = require('base32.js')
+const parallel = require('async/parallel')
+const mapLimit = require('async/mapLimit')
+const { Key } = require('interface-datastore')
+const log = require('debug')('ipfs:gc')
+
+// Limit on the number of parallel block remove operations
+const BLOCK_RM_CONCURRENCY = 256
+const MFS_ROOT_DS_KEY = new Key('/local/filesroot')
+
+// Perform mark and sweep garbage collection
+module.exports = function gc (self) {
+  return promisify((callback) => {
+    const start = Date.now()
+    log(`Creating set of marked blocks`)
+
+    self._gcLock.writeLock((lockCb) => {
+      parallel([
+        // Get all blocks from the blockstore
+        (cb) => self._repo.blocks.query({ keysOnly: true }, cb),
+        // Mark all blocks that are being used
+        (cb) => createMarkedSet(self, cb)
+      ], (err, [blocks, markedSet]) => {
+        if (err) {
+          log(`Error - ${err.message}`)
+          return lockCb(err)
+        }
+
+        // Delete blocks that are not being used
+        deleteUnmarkedBlocks(self, markedSet, blocks, start, (err, res) => {
+          if (err) {
+            log(`Error - ${err.message}`)
+            return lockCb(err)
+          }
+          lockCb(null, res)
+        })
+      })
+    }, callback)
+  })
+}
+
+// Get Set of CIDs of blocks to keep
+function createMarkedSet (ipfs, callback) {
+  parallel([
+    // All pins, direct and indirect
+    (cb) => ipfs.pin.ls((err, pins) => {
+      if (err) {
+        return cb(new Error(`Could not list pinned blocks: ${err.message}`))
+      }
+      log(`Found ${pins.length} pinned blocks`)
+      const cids = pins.map(p => new CID(p.hash))
+      // log('  ' + cids.join('\n  '))
+      cb(null, cids)
+    }),
+
+    // Blocks used internally by the pinner
+    (cb) => ipfs.pin._getInternalBlocks((err, cids) => {
+      if (err) {
+        return cb(new Error(`Could not list pinner internal blocks: ${err.message}`))
+      }
+      log(`Found ${cids.length} pinner internal blocks`)
+      // log('  ' + cids.join('\n  '))
+      cb(null, cids)
+    }),
+
+    // The MFS root and all its descendants
+    (cb) => ipfs._repo.root.get(MFS_ROOT_DS_KEY, (err, mh) => {
+      if (err) {
+        if (err.code === 'ERR_NOT_FOUND') {
+          log(`No blocks in MFS`)
+          return cb(null, [])
+        }
+        return cb(new Error(`Could not get MFS root from datastore: ${err.message}`))
```

Comment:
It would be nice not to lose the stack from this error! There's quite a lot of instances in this PR where this is happening and it'll make debugging really difficult.

If you want to add context to the error then I'd recommend using [`explain-error`](https://www.npmjs.com/package/explain-error).

Reply:
I wanted to add context for a user who gets something like 'ENOENT file descriptor invalid' and has no idea what it means. explain-err looks like a great way to help with this 👍